extensions/amp-skimlinks/0.1/tracking.js

```diff
@@ -0,0 +1,290 @@
+   * register page impression request.
+   * @param {!Object} commonData
+   * @param {number} numberAffiliateLinks
+   * @private
+   */
+  sendPageImpressionTracking_(commonData, numberAffiliateLinks) {
+    const {customTrackingId, referrer} = this.trackingInfo_;
+
+    const data = /** @type {!JsonObject} */ (Object.assign(
+        dict({
+          'slc': numberAffiliateLinks,
+          'jsl': 0, // Javascript load time, not relevant in AMP context.
+          'pref': referrer,
+          'uc': customTrackingId,
+          't': 1,
+        }),
+        commonData
+    ));
+
+    // Sends POST request. Second param is the object used to interpolate
+    // placeholder variables defined in PAGE_IMPRESSION_TRACKING_URL.
+    this.analytics_.trigger('page-impressions', dict({
+      'data': JSON.stringify(data),
+    }));
+  }
+
+  /**
+   * Sends tracking request to Skimlinks tracking API in order to
+   * register link impression request.
+   * @param {!Object} commonData
+   * @param {number} numberAffiliateLinks
+   * @param {!JsonObject} urls
+   * @private
+   */
+  sendLinkImpressionTracking_(commonData, numberAffiliateLinks, urls) {
+    const data = /** @type {!JsonObject} */ (Object.assign(
+        dict({
+          'dl': urls,
+          'hae': numberAffiliateLinks ? 1 : 0, // 1 if has at least one AE link
+          'typ': 'l',
+        }),
+        commonData
+    ));
+
+    // Send POST request. Second param is the object used to interpolate
+    // placeholder variables defined in LINKS_IMPRESSIONS_TRACKING_URL.
+    this.analytics_.trigger('link-impressions', dict({
+      'data': JSON.stringify(data),
+    }));
+  }
+
+  /**
+   * Initialise the amp-analytics internal API.
+   * Warning: Analytics API will not send any request until
+   * CommonSignals.LOAD_START has been triggered.
+   * (See "initTracking_" in amp-skimlinks.js)
+   * @param {!AmpElement} element
+   * @return {!Object}
+   * @private
+   */
+  setupAnalytics_(element) {
+    const analyticsBuilder = new CustomEventReporterBuilder(element);
+    // Configure analytics to send POST request when receiving
+    // 'page-impressions' event.
+    analyticsBuilder.track('page-impressions', PAGE_IMPRESSION_TRACKING_URL);
+    analyticsBuilder.track('link-impressions', LINKS_IMPRESSIONS_TRACKING_URL);
+    analyticsBuilder.track('non-affiliate-click', NA_CLICK_TRACKING_URL);
+
+    const analytics = analyticsBuilder.build();
+    /*
+      Use {beacon: true} to send the request through sendBeacon()
+      https://www.ampproject.org/docs/analytics/deep_dive_analytics#how-data-gets-sent:-transport-attribute
+
+      Overwrite config manually since CustomEventReporterBuilder doesn't
+      support optional config.
+      TODO: add optional config param to .build() so we don't need to mutate
+      a private property from outside.
+    */
+    // Bracket because analytics.config_ is of type JsonObject.
+    analytics.config_['transport'] = dict({'beacon': true});
```

Comment:
This won't work, the access to private variable will be broken after compilation. 

@slocka To confirm: you only want to send request via `sendBeacon()`? what about the case when beacon is not supported? 
By default the transport config will be `{image: true, post: true, beacon: true}` will that work for you?

opened #19014

Reply:
@zhouyx Is there an alternative solution until the fix for https://github.com/ampproject/amphtml/issues/19014 is ready?

We want to use sendBeacon when possible since some of the requests will be sent when the user clicks on an external link and we want to make sure the request is sent before the browser navigates away. I believe using an image would work as well. `{'beacon': true, 'image': true}` is probably the best option then. However I am not familiar with the amp-analytics behavior, is the triggered event handled asynchronously? Is there a chance that the `navigator.sendBeacon()` doesn't get called in time?

Reply:
@slocka I think the default transport config will work for you if you send request sent using xhr.
The logic is as following.
Use sendBeacon if that's possible. If browser doesn't support that fallback to sending request using xhr. If that's not supported fallback to using image.

The other question: The triggered event is handled synchronously as long as you don't use batching or have the event wait for promise to be resolved.

Reply:
ping @zhouyx

Reply:
@slocka Sorry for the delay, was on vacation. Thank you @lannka for the review.